test/object-store/frozen_objects.cpp

```diff
@@ -133,29 +133,63 @@ TEST_CASE("Freeze Results", "[freeze_results]") {
-            child_obj.set(linked_object_value_col, j + 10);
-            object_link_view->add(child_obj.get_key());
+            if (with_links) {
+                auto child_obj = linked_table->create_object();
+                child_obj.set(linked_object_value_col, j + 10);
+                object_link_view->add(child_obj.get_key());
+            }
             int_list.add(static_cast<Int>(j + 42));
             std::string key = "Key" + util::to_string(j);
-            int_dict.insert(key, i);
+            int_dict.insert(key, value);
         }
-    }
-    realm->commit_transaction();
+        return obj;
+    };
+
+    auto write = [&](const std::function<void()>& fn) {
+        realm->begin_transaction();
+        fn();
+        realm->commit_transaction();
+    };
+
+    write([&]() {
+        for (int i = 0; i < 8; ++i)
+            create_object(i + 2, true);
+    });
 
-    Results results(realm, table);
     auto frozen_realm = Realm::get_frozen_realm(config, realm->read_transaction_version());
-    Results frozen_results = results.freeze(frozen_realm);
+
+#define VERIFY_VALID_RESULTS(results, realm)                                                                         \
+    REQUIRE(results.is_valid());                                                                                     \
+    if (results.size() == 0) {                                                                                       \
+        REQUIRE_EXCEPTION(results.get_any(0), OutOfBounds,                                                           \
+                          "Requested index 0 calling get_any() on Results when empty");                              \
+    }                                                                                                                \
+    else {                                                                                                           \
+        REQUIRE_FALSE(results.get_any(0).is_null());                                                                 \
+    }                                                                                                                \
+    REQUIRE(results.freeze(realm).is_valid());
+
+#define VERIFY_INVALID_RESULTS(results, realm, exception, message)                                                   \
+    REQUIRE_EXCEPTION(results.freeze(realm), exception, message);                                                    \
+    REQUIRE_FALSE(results.is_valid());                                                                               \
+    REQUIRE_EXCEPTION(results.size(), exception, message);                                                           \
+    REQUIRE_EXCEPTION(results.get_any(0).is_null(), exception, message);
+
+#define VERIFY_STALE_RESULTS(results, realm)                                                                         \
+    VERIFY_INVALID_RESULTS(results, realm, StaleAccessor, "Access to invalidated Results objects")
```

Comment:
Type of exception will change for Results on table and query from realm::NoSuchTable. But it will still crash eventually on freeze, so not a big deal, right?